src/app/lib/recipes/parser.ts

```diff
@@ -0,0 +1,65 @@
+import {BrewingRecipe, Recipe} from "@/app/lib/definitions";
+import {roundAndTruncate} from "@/app/ui/utils";
+
+
+export const getBrewingRecipeUsingTotalWater = (rawRecipe: Recipe, totalWaterAmount: number): BrewingRecipe => {
+  let coffeeAmount = (rawRecipe.coffeeGramsPerLiter * totalWaterAmount) / 1000
+  return getBrewingRecipe(rawRecipe, totalWaterAmount, coffeeAmount)
+}
+
+export const getBrewingRecipeUsingCoffeeAmount = (rawRecipe: Recipe, coffeeAmount: number): BrewingRecipe => {
+  let totalWaterAmount = (coffeeAmount * 1000) / rawRecipe.coffeeGramsPerLiter
+  return getBrewingRecipe(rawRecipe, totalWaterAmount, coffeeAmount)
+}
+
+
+const evalRecipeSchemaTemplateString = (
+  templateString: string | undefined,
+  stepWaterAmountRaw: number,
+  waterAccumulatorRaw: number,
+  stepDuration: number
+): string => {
+  if (templateString == null) {
+    return ""
+  }
+  let flowRate = roundAndTruncate(stepWaterAmountRaw / stepDuration, 1)
+  let stepWaterAmount = roundAndTruncate(stepWaterAmountRaw, 1)
+  let waterAccumulator = roundAndTruncate(waterAccumulatorRaw, 1)
+
+  let templateToEval = "`" + templateString.replaceAll("{", "${") + "`"
+  return eval(templateToEval)
```

Comment:
this is horrible, but easiest. we need to do manual replaces probably with regex, for now I would stick to that, it works kinda fine, and vulnerability takes place only in recipe definition, so this is probably secure for now. Also Wiktoria suggested that there might be multiple libraries which does just that in more secure way, so I will create a ticket to investigate and fix that

Reply:
https://github.com/users/jakub-filistynski/projects/3/views/1?pane=issue&itemId=60583376